p2p/server.py

```diff
@@ -122,36 +143,121 @@ def __init__(self,
         )
 
         # Create and register peer in peer_pool
-        eth_peer = ETHPeer(
-            remote=initiator_remote, privkey=self.privkey, reader=reader,
-            writer=writer, aes_secret=aes_secret, mac_secret=mac_secret,
-            egress_mac=egress_mac, ingress_mac=ingress_mac, chaindb=self.chaindb,
+        peer = self.peer_class(
+            remote=initiator_remote,
+            privkey=self.privkey,
+            reader=reader,
+            writer=writer,
+            aes_secret=aes_secret,
+            mac_secret=mac_secret,
+            egress_mac=egress_mac,
+            ingress_mac=ingress_mac,
+            chaindb=self.chaindb,
             network_id=self.network_id
         )
-        self.peer_pool.add_peer(eth_peer)
-
-
-def decode_authentication(ciphertext: bytes,
-                          privkey: datatypes.PrivateKey
-                          ) -> Tuple[datatypes.PublicKey, bytes, datatypes.PublicKey]:
-    """
-    Decrypts and decodes the ciphertext msg.
-    Returns the initiator's ephemeral pubkey, nonce, and pubkey.
-    """
-    if len(ciphertext) < ENCRYPTED_AUTH_MSG_LEN:
-        raise ValueError("Auth msg too short: {}".format(len(ciphertext)))
-    elif len(ciphertext) == ENCRYPTED_AUTH_MSG_LEN:
-        sig, initiator_pubkey, initiator_nonce, _ = decode_auth_plain(
-            ciphertext, privkey)
+
+        await self.apply_p2p_handshake(peer)
+
+    async def apply_p2p_handshake(self, peer: BasePeer) -> None:
+        finish_p2p_handshake = False
+        try:
+            # P2P Handshake.
+            await wait_with_token(
+                peer.do_p2p_handshake(),
+                token=peer.cancel_token,
+                timeout=HANDSHAKE_TIMEOUT,
```

Comment:
This is a rather long timeout as it is intended to be used as the overall timeout for both the auth and p2p handshakes.  I think the best way forward is to not use `wait_with_token()` here, rename the current `receive_handshake()` to `_receive_handshake()` and create a new `receive_handshake()` like this:
```
async def receive_handshake(...):
    return wait_with_token(self._receive_handshake(...), token=self.cancel_token, timeout=HANDSHAKE_TIMEOUT)
```

Also, while you're at it, the `reader.read()` calls in `receive_handshake()` need to be wrapped with `wait_with_token()` as well

Reply:
@gsalgado  Thank you so much! For `reader.read()`, I added `wait_with_token` with `timeout=REPLY_TIMEOUT` (p.s. `HANDSHAKE_TIMEOUT = 5`, `REPLY_TIMEOUT = 3`), does that make sense?